### ISSUE TITLE
knock back on entities by tnt ignores weight

### ISSUE THREAD

**Issue by DS-Minetest** [minetest/minetest_game#1300]
in object property table you can set a weight [(api)](https://github.com/minetest/minetest/blob/master/doc/lua_api.txt#L3455)
but - please correct me if im false - the weight is ignored by the impulse from the exploding tnt

p = m*v; v = p/m; => v ~ m⁻¹

**Comment by sofar** [comment]
This could be added, but it makes it a lot more complex, plus it would look weird, I think.

Inclined to say that the simplification we have now is preferred.

**Comment by paramat** [comment]
I iagree with sofar.

**Comment by sofar** [comment]
Elaborating on why this is somewhat a bad idea:

If we'd only use `weight` to calculate the impulse on objects, then we'd see that for light things would get immense speeds, and heavy things would get blown only a small distance. It's quite likely that most things in the world are light, but things like blocks of sand and stone, usually around, are likely to be very heavy, after all a cubic meter of stone is at minimum 2400kg (5klbs) or so weight.

You'd end up with a situation where you'd see stuff flying hundreds of meters, but at the same time some stuff will end up almost not moving at all due to the incredible range of weights: a shovel weighs maybe 2kg, and there's plenty of items in minetest that are far lighter than that.

Now, your physics formula listed above is entirely correct, but it omits an entire, and very important part, of the impulse generated by the TNT explosion: `area`.

The explosion shockwave creates compressed air/gas that pushes against the objects. The energy transferred in this shockwave to the object's kinetic energy is proportional to the `surface` of the object. So, a small object will have a small impulse, and a large object will get a larger impulse.

Now, we have only very limited idea about the surface area of objects. We may make a guess about the hitbox, but it completely ignores the shape. If we'd actually do the shape calculation correctly, we'd have to calculate the effect of the spherical shockwave on the particular sides of the object's hitbox or bounding box, which is surely going to involve integration and similar advanced calculus.

So, in conclusion: If we really want to incorporate `weight`, we should really incorporate `shape/size` as well - otherwise it's just very incorrect.

Mind you, most 1x1x1 blocks have a large surface, and are heavy, and most items have a light weight, and are small. So the end result is likely going to be very similar to what the code does right now since those two factors are likely largely cancelling each other out.

So in the end, it's just better to not do this.

**Comment by DS-Minetest** [comment]
I agree with sofar.

**Comment by ExeterDad** [comment]
Very good point @sofar. But rather than attempt to calculate the surface just assign a value of 1 (a full node), then when the item is registered, leave it to the modder to assign a proper surface. A slab could be .5. It might take some time for modders to catch up and updater their mods, but at least the code would be there.

**Comment by paramat** [comment]
Spot-on points by sofar.

**Comment by DS-Minetest** [comment]
you would need the Wc from each side of an object
or you ignore the shape of the surface...

**Comment by DS-Minetest** [comment]
no problem, you are right
